### ISSUE TITLE
Implementation of SWP-08 routing protocol

### ISSUE THREAD

**Issue by evenover** [bitfocus/companion-module-generic-swp08#1]
Hey!

I would like companion to implement the SWP-08 router protocol. Its a protocol supported by alot of different router-systems like ross ultrix, direct out matrix-routing and lots more.
I want to be able to perform routes and read out routing info. Then be able to use the feedback from the router to make a tallysystem.

Documentation apllies to both SWP-02 and 08
https://wwwapps.grassvalley.com/docs/Manuals/sam/Protocols%20and%20MIBs/Router%20Control%20Protocols%20SW-P-88%20Issue%204b.pdf

**Comment by evenover** [comment]
Now i have been testing a bit and here are som points:
1. How can i get the names from the router onto my buttons? should they be stored in variables?
2. I want to have a feedback that i can but on the sources that change color when the destination it is routed to is selected.
3. It should be a variable with name that is the source of the selected destination.
4. I used the "select destination" and when I choose one destination and then another, the previuosly chosen destination inherit the now selected destinations destination(complicated)

**Comment by chibiconsulting** [comment]
FYI: Although the version linked at https://we.tl/t-5nurvA8mcQ is a different size, it has the same version number as the last one of 2.2.0 (2.2.0-061e92b1-3423)

Re: level multiselect.
The observed behaviour of the buttons in the emulator do not seem to reflect multiselect. In normal button mode, each button will flash the active (yellow top bar) indication when pressed and the feed back will light up the last button pushed. In Latch/Toggle button mode, each button will light the active (yellow top bar) indication when pressed; however, it will never turn off. The feed back will still light up the last button pushed. I would guess that the buttons would want to be normal rather than using the predefined Latch/Toggle logic?

Is this behavior simply an anomaly of the emulator that is not present with an active router controller?

Also, this may be a feature request; but, I see the ability to select more than one level per action & feedback is present; would it be possible to define overlapping level buttons?

For example, having four buttons for three levels where the first three are toggle on/off and the fourth is an override to toggle all three selected on or off. Here would be the definitions for these buttons for testing.

Button name: L1
Press Action: SW-P-08: Select Levels	1
Feedback: SW-P-08: Selected Levels	1

Button name: L2
Press Action: SW-P-08: Select Levels	2
Feedback: SW-P-08: Selected Levels	2

Button name: L3
Press Action: SW-P-08: Select Levels	3
Feedback: SW-P-08: Selected Levels	3

Button name: All Levels
Press Action: SW-P-08: Select Levels	1,2,3
Feedback: SW-P-08: Selected Levels	1,2,3

**Comment by chibiconsulting** [comment]
Selected levels always need to be visible via feedback for those who need to be able to do split routes. As mentioned before, most people will only ever want simultaneous routing; however, for those who need to be able to break out Audio and Video levels, they need to see the current level status directly on the panel via feedback.

Turning off all level selection is a common operational procedure rather than a true lockout. Many router control panels are in areas on desks that can be bumped and being able to deselect levels reduces issues; however, usually, the same button is also used to select all levels again rather than to select each level individually so as to minimize the number of intentional button pushes. So a clear option is only useful with a select all option to go with it.

Re: lockouts, while SW-P-08 has a robust set of destination lock and router lock options, it is probably best not to support those at all; as those can be much more complex than routing itself, and, if someone needs that level of control they should probably be solving that problem direct with the specific router vendor. I think that using the existing lockout options within Companion is more than sufficient.

**Comment by daniep01** [comment]
I think I understand that the selection level buttons need to be additive, so that combinations can be created. This will take a bit of thought.

One thing I have done already is the make the clear action more configurable so it can clear only levels if necessary.

**Comment by chibiconsulting** [comment]
Thanks.
I also notice that there is an additional method to get feedback beyond the way I mentioned above.
Feedback:  internal: Check variable value SW-P-08:Level - Selected levels = []
That method has available operators of [ =, !=, >, and < ]; however, does not have [AND] [OR] options so it is not clear how that would support multi level feedback.
The Feedback: SW-P-08: Selected Levels method on the other hand seems to support multiple levels already so that may not need any further changes.

**Comment by daniep01** [comment]
I currently have 2 modes of operation. Firstly:

- level select and deselct which can take multiple levels
- source select
- destination select
- take or clear

 This makes one 'set' of actions. Then we also have:

- crosspoint a level/source/destination in a single action

I'm wondering if this second mode takes the level selection in the first method or ignores it?

**Comment by chibiconsulting** [comment]
My vote would be to keep the operational modes separate if possible. @evenover do you have a preference? If a single button push is defined, it should have the ability to have control over all aspects of that action regardless of what state the control surface was previously in.

For the other mode of operation, the default mode should be that all levels are selected at startup or after a Clear command. The ability to split levels is a more advanced method and should only be enabled if explicitly requested. For example, if I walk up to a control surface and do not know what the previous condition was, this is the typical expected sequence.

- Hit Clear. {Reset to default condition.}
- Select a Destination. {This allows the ability to see what is currently routed there.}
- Select a new Source.
- Hit the Take button. {Modifying the levels is optional and can be changed at any point before the Take.}

This should result in all of levels of the selected source being routed to that destination. If the levels are changed for a route those can then persist until cleared or manually changed. BTW: Some control systems implement a timeout function to Clear the routing presets after a period in case someone has a complex route setup and then walks away, which could cause unexpected results for the next user.

**Comment by evenover** [comment]
It seems to be a good idea @chibiconsulting. A timeout-function would also be great.
The way i am planing to use this module is to have kind of a catergory select. So i'm going to build separate pages for Video, audio and then video+audio(video+8 audio probably). This way i am not going to make use of turning levels on and off but just add level-selection to the button with the destination. 
I think this protocol actually only supports routing 1 level at the time, thats probably why its hard to get feedback of multiple levels selected.

I have managed to obtain the latest full documentation of swp08 with extended routing/labels from Grass Valley, but i have to check that my employer do not have a NDA with Grass before i can supply it to you. Hopefully its not a big deal and i will give it to you as soon as i know.

**Comment by daniep01** [comment]
@evenover It's great that you have the documentation but remember that companion and this module is open source so GV need to be aware that it's trivial to reverse engineer their spec from the code in this module.

**Comment by daniep01** [comment]
Hi guys, in order to get this module out to more potential users it has been included in the next beta of companion as the basic routing functionality is all working. This doesn't mean I won't make the more complex changes requested above.

**Comment by evenover** [comment]
Thats great!

Any updates on the features:
1. Change color on background of source of destination when destination is selected
2. variables for label names which can be added to buttons
3. variable with the source name of selected destination

1 and 2 are pretty much the same features as the videohub-module.

I took a look in the manual for ultrix and it only supports 8-char labels, but 16char umds of the sources. In the extended documentation its possible to ask for 4, 8 or 12. Hopefully i can disclose it with you asap.

**Comment by daniep01** [comment]
I've redone the level select/deselect actions so they work as toggles, next step is to make feedbacks work for those and then the list above.

**Comment by daniep01** [comment]
The level commands have been rewritten to work as follows:

- Select Level: add the level(s) in the action to the current selection
- DeSelect Level: remove the level(s) in the action from the current selection
- Feedback: Change button colour if **all** the level(s) in the feedback are in the current selection

You would normally put all of the above on one phyiscal button and enable latch/toggle mode. The clear action re-enables all levels up to the number set in the module configuration.

**Due to all the changes you will need to delete any previous level actions and feedbacks from your configuration before trying this!**

https://we.tl/t-RqZEdtxs3c

**Comment by daniep01** [comment]
Having done more testing the above method can sometimes get in a mess when the latch state gets out of step with the level state. To avoid this there's a new action which will simply toggle (invert) the level(s) in the action each time it is pressed. This is much preferable to using the button latch mode. 

https://we.tl/t-LmLxyBaQ26

**Comment by chibiconsulting** [comment]
I like the SW-P-08: Toggle Levels action, nice. Can still see some people using the SW-P-08: Select Levels and SW-P-08: De-Select Levels actions, hope those stay as well.

**Comment by chibiconsulting** [comment]
@evenover 

> 1. How can i get the names from the router onto my buttons? should they be stored in variables?

- I would think that if available by feedback it would be enough, or do you have something else in mind?

i.e. Feedback: SW-P-08: Name of target. or Name of Source and Name of Destination for example.

> 2./3. I want to have a feedback that i can put on the sources that change color when the destination it is routed to is selected.

- Something like Feedback: SW-P-08: Source routed to Destination? Would need to query the selected destination for the current Source. What happens if the routed Source changes after the button is updated? Could use a lot of cycles to keep checking.

> 4. I used the "select destination" and when I choose one destination and then another, the previously chosen destination inherit the now selected destinations destination(complicated)

- I think that Evertz implements this as a virtual destination which is a destination that can then be used as a source somewhere else. I do not recall SW-P-08 ever having that capability. Might make sense as a separate, more generic request to provide that function for all router protocols. If a range of virtual sources/destinations are defined as being linked and held separate from the router, requests to / from those can be translated on the fly and dropped back to the native router protocol. Not a trivial problem by any means.

**Comment by daniep01** [comment]
> > How can i get the names from the router onto my buttons? should they be stored in variables?
> 
> I would think that if available by feedback it would be enough, or do you have something else in mind?
> i.e. Feedback: SW-P-08: Name of target. or Name of Source and Name of Destination for example.

Feedbacks are really intended for modifying the button colour or text, I think what is being asked here is that the button label can be set from a variable, please confirm.

> 
> > I want to have a feedback that i can put on the sources that change color when the destination it is routed to is selected.
> 
> Something like Feedback: SW-P-08: Source routed to Destination? Would need to query the selected destination for the current Source. What happens if the routed Source changes after the button is updated? Could use a lot of cycles to keep checking.

Currently the router sends a message every time a route changes (look in the debug log to see these) so I need to write some code to keep track of all these messages. What makes it complicated compared to say the videohub is keeping track of all the potential breakaway routes.
> 
> > I used the "select destination" and when I choose one destination and then another, the previously chosen destination inherit the now selected destinations destination(complicated)
> 
> * I think that Evertz implements this as a virtual destination which is a destination that can then be used as a source somewhere else. I do not recall SW-P-08 ever having that capability. Might make sense as a separate, more generic request to provide that function for all router protocols. If a range of virtual sources/destinations are defined as being linked and held separate from the router, requests to / from those can be translated on the fly and dropped back to the native router protocol. Not a trivial problem by any means.

This is a quite advanced feature and in my opinion companion is generally aimed at casual or supplementary use cases. For this kind of complexity you should probably be using the Ross panels or similar.

**Comment by daniep01** [comment]
Now this module has a repository we can log all these requests separately and this thread is getting too long. I have transferred the feature requests I am aware of to new issues. Please check and update as required. Please log new bugs and features separately.

**Comment by scp85** [comment]
I can't find the repository, could you tell me where to find it? All the wetransfer links are obviously expired.
Thanks

> Now this module has a repository we can log all these requests separately and this thread is getting too long. I have transferred the feature requests I am aware of to new issues. Please check and update as required. Please log new bugs and features separately.

**Comment by daniep01** [comment]
@scp85 The latest build is in the beta companion app - [https://bitfocus.io/companion/download/builds/](https://bitfocus.io/companion/download/builds/)

**Comment by karthickvkl** [comment]
is there any IP based emulator available for SWP-08 routing protocol?

**Comment by evenover** [comment]
> is there any IP based emulator available for SWP-08 routing protocol?

Not that I'm aware of.
What were you thinking using it for?

**Comment by scp85** [comment]
I’m trying to figure out ways to control an Swp-08 audio/video router (grassvalley (used to be Snell Andvanced Media) Sirius S830)) with a PC/ audio desk..

**Comment by evenover** [comment]
This module should be working fine. I've done some testing on my Ross Ultrix and it looks good for now. If you have over 1000 sources and destinations there are still some developments that need to happen.

**Comment by scp85** [comment]
Ouch! 8000 I/o for me.
I’ll still give it a go tho

**Comment by evenover** [comment]
If we get the extended comands implemented it will be able to control 65536 sources and destinations.

**Comment by daniep01** [comment]
I'm afraid that because I don't require the extended commands in my use cases the development time would need to be paid for somehow.

**Comment by evenover** [comment]
I understand, any idea on what kind of payment? I might be able to fix some funding.

**Comment by daniep01** [comment]
I think £200 would cover it.

**Comment by evenover** [comment]
Awesome, that should be no problem!

Would it then be possible to see if you could get the "change background color of selected source" working?

**Comment by daniep01** [comment]
Maybe... let's do the extended commands first. Please email me to discuss payment mail-at-pjdaniel-dot-org-dot-uk

**Comment by InertiaImpact** [comment]
Any chance you would know if this is able to work with a nVision NV9000 Router controller from Grass Valley?

**Comment by engelCZE** [comment]
> Any chance you would know if this is able to work with a nVision NV9000 Router controller from Grass Valley?

Hi, I think it could be working, NV9000 is on swp02. This is router protocol swp08 is remote protocol and we are using swp08 for evertz switcher control thru CEREBRUM automatization via swp08

**Comment by InertiaImpact** [comment]
Thanks, hmmmm now I just have to figure out how to set that up lol

**Comment by mplakalovi1** [comment]
Hello team, I'm curious to know whether it's possible to disconnect a crosspoint using the _**Pro-Bel SW-P-08**_ protocol. _**Can a crosspoint connection be directly removed, or is the standard procedure to always redirect the input to a specific output?**_ 

Moreover, is there a way to work-around this behavior? For example, could we route the input to a designated output that can be deactivated on the physical device?

Thank you very much in advance !

**Comment by chibiconsulting** [comment]
No. This is not specific to SW-P-8 control, all routers have this constraint.
For Video or Audio routers, the standard practice is to route a specific input to any outputs that no longer require the previous input. Examples would be a blank screen and silent audio or bars and tone. Most systems already have a source for those anyway.

This is due to the physics of the crosspoint switching mechanism and associated amplification within the device. To put it another way, an unterminated source being output by a router would cause undesirable operation. ;)

Note that data routers are an exception, where the router itself should define a disconnect source and you would need to check the specifics with the data sheet for those devices; however, from a protocol perspective, it still places a crosspoint select message using whatever the disconnect source is defined as.

**Comment by mplakalovi1** [comment]
Thank you @chibiconsulting for your prompt and informative response; I was seeking for anything like this... 

All uncertainties are resolved!

**Comment by Gbou7** [comment]
> Thanks, hmmmm now I just have to figure out how to set that up lol

I'm wandering if you managed to get it to work ? Because I'm using a NV9000 too but it's not working for me for now
Thanks

**Comment by engelCZE** [comment]
> > Thanks, hmmmm now I just have to figure out how to set that up lol
> 
> 
> 
> I'm wandering if you managed to get it to work ? Because I'm using a NV9000 too but it's not working for me for now
> 
> Thanks

Hi we're using SW-P-08 generic and works for us, just fill IP and port what we are using and it works. Truth is that we are using semi-software Cerebrum for manage control of matrix. Means Elgato->swp08->Cerebrum->swp08->Evertz

**Comment by willosof** [comment]
@daniep01 @evenover is this module in good shape now? I'll arrange funding if it can be improved in some way.

**Comment by phillipivan** [comment]
Hi William,
I can probably speak to that reasonably having done a bit of work on the module recently. It works reasonably but the implementation is not very complete. A few obvious areas for improvement:

- Request and process Tally Dump
- Better multilevel crosspoint tallies
- Protect / Extended Protect Set / Tally / Request Tally Dump

One of the potential pitfalls with SW P 08 is that there is a large number of commands specified, but few if any devices support all of them. I faced a similar issue with the SW P 02, and only implemented the scope I could test.

For instance [this](https://www.directout.eu/wp-content/uploads/m1k2/www_ship/en/help/swp08.html) is the portions of SW P 08 that a specific Direct Out device supports. Perhaps a quarter of all commands listed, which is probably not complete to begin with.

**Comment by Corent1CSN-LAB** [comment]
Hello,

I’m experiencing issues with the SWP-08 and the Ultrix FR-12 Frame. While I can send control commands, I’m not receiving any feedback or information from the equipment.

Has anyone worked with Ross Ultrix systems and encountered a similar issue? Any troubleshooting suggestions would be greatly appreciated.

Thanks!

**Comment by phillipivan** [comment]
Yep, we use it with a smaller Ultrix. I can have a look at the Ultrix side config next week and see if there is anything notable we had to do.

Can you please check the module logs for return data, and let me know what you find.

**Comment by pomerle** [comment]
Hi Guys, Im a solutions architect with the Hyerconverged team at Ross. The FR12 uses the Ultricore BCS as a Control system. Make sure you are sending your commands to the Ultricore BCS and not to the frame directly. Also it is possible that the probel service is not enabled. We implimented new security features that disables most services by default now (version 5.5 and up). On the BCS under product info, services. 

I've tested this today with Companion 3.4.3 and Ultricore BCS / Ultrix  version 5.6 and 6.2 (with an FR12, FR5, FR2)  and it works as expected.

**Comment by Corent1CSN-LAB** [comment]
> Hi Guys, Im a solutions architect with the Hyerconverged team at Ross. The FR12 uses the Ultricore BCS as a Control system. Make sure you are sending your commands to the Ultricore BCS and not to the frame directly. Also it is possible that the probel service is not enabled. We implimented new security features that disables most services by default now (version 5.5 and up). On the BCS under product info, services.
> 
> I've tested this today with Companion 3.4.3 and Ultricore BCS / Ultrix version 5.6 and 6.2 (with an FR12, FR5, FR2) and it works as expected.

Hello, indeed my issue was there; I was not pointing to the virtual IP address of the cluster. The grid now appears perfectly in the module. Thank you!
